# ■ 회귀분석
    1. 단순 회귀분석
        - 단순 회귀분석 이론 
        - 단순 회귀분석 실습1 : 애벌레 사료와 탄닌 함유량과의 관계
        - 단순 회귀분석 실습2 : 우주 왕복선 챌린저 폭파원인
        - 상관관계 ( 다중회귀 -> 다중공선성(multicolinearity)
        - 자동화 스크립트 개선 ( 산포도 그래프 + 회귀직선 추가 )            
    2. 다중 회귀분석
        

## ■ 다중회귀분석
```
    단순선형회귀분석 -> 목적이 하나의 독립변수만을 가지고 종속변수를 예측하기 위한 회귀모형을 만들기 위한 것
    다중회귀분석 -> 목적이 여러개의 독립변수들을 가지고 종속변수를 예측하기 위한 회귀모형을 만들기 위한 것
```
```
    질문1. 집 값에 영향을 미치는 요소가 여러 독립 변수들 중 무엇?
```

![다중회귀식](http://cfile259.uf.daum.net/image/9923BA3E5EF897C425BDC1)

### □ 다중선형 회귀식을 이해하기 위한 사전 지식
```
        1. 전치행렬 : 행과 열을 교환하여 얻은 행렬
        2. 단위행렬 : 주 대각선의 원소가 모두 1이며 나머지 원소는 모두 0인 정사각 행렬
        3. 역행렬 : 원래 행렬과 곱했을때 단위행렬이 되는 행렬
```
![역행렬](http://cfile263.uf.daum.net/image/99DB88335A0BB5441C52A3)
```
        4. 의사역행렬(Pseudo inverse matrix) 
            역행렬은 정방행렬일 때만 구할 수 있다. 따라서 직사각행렬일 때는 역행렬을 구할 수 없으므로 의사역행렬을 구해야한다
            - 정방행렬
```
$$B=\begin{pmatrix} a & b \\  c & d \\  \end{pmatrix}$$
$$B^{-1}={{1}\over{ad-bc}}\begin{pmatrix} d & -b \\ -c & a \\ \end{pmatrix}$$
```
            - 비 정방행렬
```
$$A=\begin{pmatrix} x_{11} & x_{12} & x_{13} & x_{14} \\ x_{21} & x_{22} & x_{23} & x_{24} \\ x_{31} & x_{32} & x_{33} & x_{34} \\ \end{pmatrix}$$
$$A \cdot A^{T} : 정방행렬　　det(A)≠0　일　때　역행렬이　존재$$

### ※ 문제255. 아래의 행렬을 r로 구현하고 아래의 행렬의 전치행렬을 구하시오
$$ \begin{pmatrix}
    1 & 2 \\
    3 & 4 \\
    5 & 6 \\
    \end{pmatrix}$$

In [1]:
A <- matrix(c(1,2,3,4,5,6), nrow = 3, ncol = 2, byrow=T)
A
t(A)

[,1] [,2] [,3]
[1,] 1    2    3   
[2,] 4    5    6

### ※ 문제256. 아래의 행렬의 곱을 구현하시오
$$ \begin{pmatrix}
    1 & 2 & 3 \\
    2 & 3 & 4 \\
    \end{pmatrix}*
    \begin{pmatrix}
    1 & 0 & 0 \\
    0 & 1 & 0 \\
    0 & 0 & 1 \\
    \end{pmatrix} $$

In [7]:
b <- matrix(c(1,2,3,2,3,4),nrow=2,ncol=3, byrow=T)
i <- diag(3)
b%*%i

[,1] [,2] [,3]
[1,] 1    2    3   
[2,] 2    3    4

### ※ 문제257. 역행렬구하기
$$\begin{pmatrix}
    1 & 2 \\
    3 & 4 \\
    \end{pmatrix}
    \begin{pmatrix}
    x_{11} & x_{12} \\
    x_{21} & x_{22} \\
    \end{pmatrix}=\begin{pmatrix}
    1 & 0 \\
    0 & 1 \\
    \end{pmatrix}$$

In [8]:
A <- matrix(c(1,2,3,4),nrow=2,ncol=2, byrow=T)
solve(A)
round(A %*% solve(A))

[,1] [,2]
[1,] -2.0  1.0
[2,]  1.5 -0.5

[,1] [,2]
[1,] 1    0   
[2,] 0    1

### ※ 문제258. 의사역행렬로 구한 다중 회귀식의 회귀 계수를 구하는 함수를 책 264페이지를 보고 구현하시오
$$ Y= X\beta + \epsilon,　　　　　 \hat{\beta}=(X^{T}X)^{-1}X^{T}Y$$

In [12]:
reg <- function(y, x){
    x <- as.matrix(x)
    x <- cbind(intercept=1,x)
    b <- solve(t(x)%*%x)%*%t(x)%*%y
    colnames(b) <- 'estimate'
    print(b)
}

    단순회귀분석 예시 : 챌린저호의 폭파원인 -> 온도, O형 링의 파손수 만 사용해서 분석
    다중회귀분석 예시 :          ~         -> 비행기번호, 압력 다 사용해서 분석

### ※ 문제259. O형링 폭파 원인에 가장 영향을 미친 요소가 온도, 비행기 번호, 압력인지 위에서 만든 reg 함수로 알아내시오

In [13]:
launch <- read.csv('challenger.csv',head=T)
lnch2 <- launch[,-1]
reg(y=launch$distress_ct,lnch2)

                         estimate
intercept             3.527093383
temperature          -0.051385940
field_check_pressure  0.001757009
flight_num            0.014292843


```
    O형링 파손에 영향을 주는 가장 큰 독립변수 : temperature -0.051385940
    그 다음 변수 : flight_num 0.014292843
```
$$ \hat{y}=3.527093383 - 0.051385940\cdot X_{1} + 0.001577009\cdot X_{2} + 0.014292843\cdot X_{3} $$
```
    온도만 가지고 단순회귀식을 도출해보면
```
$$ \hat{y}=3.70 - 0.048\cdot x_{1}$$
```
     ea          temp(Ｆ)
    2.21          31Ｆ
    0.82          60Ｆ
    0.34          70Ｆ
```
    화씨 30도에서 발사하는 것이 화씨 60도에서 발사하는 것 보다 3배 더 위험
    화씨 70도에서 발사하는 것보다 7-8배 위험